utilities/src/fixed_128.rs

```diff
@@ -0,0 +1,533 @@
+use codec::{Decode, Encode};
+use primitives::U256;
+use rstd::{
+	convert::{Into, TryFrom, TryInto},
+	num::NonZeroI128,
+};
+use sp_runtime::{
+	traits::{Bounded, Saturating, UniqueSaturatedInto},
+	PerThing,
+};
+
+#[cfg(feature = "std")]
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+use sp_runtime::traits::SaturatedConversion;
+
+/// A signed fixed-point number. Can hold any value in the range [-170_141_183_460_469_231_731, 170_141_183_460_469_231_731]
+/// with fixed-point accuracy of 10 ** 18.
+#[derive(Encode, Decode, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Fixed128(i128);
+
+const DIV: i128 = 1_000_000_000_000_000_000;
+
+impl Fixed128 {
+	/// Create self from a natural number.
+	///
+	/// Note that this might be lossy.
+	pub fn from_natural(int: i128) -> Self {
+		Self(int.saturating_mul(DIV))
+	}
+
+	/// Accuracy of `Fixed128`.
+	pub const fn accuracy() -> i128 {
+		DIV
+	}
+
+	/// Raw constructor. Equal to `parts / DIV`.
+	pub fn from_parts(parts: i128) -> Self {
+		Self(parts)
+	}
+
+	/// Creates self from a rational number. Equal to `n/d`.
+	///
+	/// Note that this might be lossy.
+	pub fn from_rational<N: UniqueSaturatedInto<i128>>(n: N, d: NonZeroI128) -> Self {
```

Comment:
It might be better to keep using `U256` for calculating `parts`, as `i128` bound is easy to be saturated into a much smaller value. Both `Fixed64` and `FixedU128` uses bigger bound ints to calculate `parts`.